### PR TITLE
ICU-22520 Replace uprv_malloc() / uprv_free() with icu::LocalPointer

### PR DESCRIPTION
**Pull request by roubert**
##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22520
- [X] Required: The PR title must be prefixed with a JIRA Issue number.
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number.
- [X] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

File: icu4c/source/common/uloc_tag.cpp [hunk 1]
```diff
@@ -858,15 +858,15 @@ _addVariantToList(VariantListEntry **first, VariantListEntry *var) {
...
-                break;
+                return false;
             }
             prev = cur;
             cur = cur->next;
         }
     }
 
-    return bAdded;
+    var.orphan();
```
Comment (markusicu) [inline icu4c/source/common/uloc_tag.cpp:868]: I think it's cleaner if you replace the var.getAlias() with var.orphan() and remove this line here.

File: icu4c/source/common/uloc_tag.cpp [hunk 2]
```diff
@@ -2386,20 +2371,21 @@ ultag_parse(const char* tag, int32_t tagLen, int32_t* parsedLen, UErrorCode* sta
         break;
     }
 
-    if (pExtension != nullptr) {
+    if (!pExtension.isNull()) {
```
Comment (markusicu) [inline icu4c/source/common/uloc_tag.cpp:2374]:
```suggestion
    if (pExtension.isValid()) {
```

File: icu4c/source/common/uloc_tag.cpp [hunk 3]
```diff
@@ -2242,11 +2231,10 @@ ultag_parse(const char* tag, int32_t tagLen, int32_t* parsedLen, UErrorCode* sta
         }
         if (next & EXTS) {
             if (_isExtensionSingleton(pSubtag, subtagLen)) {
-                if (pExtension != nullptr) {
+                if (!pExtension.isNull()) {
```
Comment (markusicu) [inline icu4c/source/common/uloc_tag.cpp:2234]:
```suggestion
                if (pExtension.isValid()) {
```

File: icu4c/source/common/uloc_tag.cpp [hunk 4]
```diff
@@ -2305,26 +2292,24 @@ ultag_parse(const char* tag, int32_t tagLen, int32_t* parsedLen, UErrorCode* sta
             if (uprv_tolower(*pSubtag) == PRIVATEUSE && subtagLen == 1) {
                 char *pPrivuseVal;
 
-                if (pExtension != nullptr) {
+                if (!pExtension.isNull()) {
```
Comment (markusicu) [inline icu4c/source/common/uloc_tag.cpp:2295]:
```suggestion
                if (pExtension.isValid()) {
```

File: icu4c/source/common/uloc_tag.cpp (hunk 1 above)
Comment (roubert) [reply icu4c/source/common/uloc_tag.cpp:868]: Done.

**Review by markusicu — APPROVED** [review]

**Pull request merged**